_posts/2024-10-24-why-new-rtos.markdown

```diff
@@ -0,0 +1,217 @@
+A futex is a conceptually simple mechanism that allows a thread to block if a memory location contains an expected value and another thread to wake one or more waiters.
+You can use a futex to build a variety of lock types and you can also use it for producer and consumer counters in a ring buffer so that producers can sleep if the queue is full and consumers can sleep if it is empty.
+
+CHERIoT RTOS implements a message queue in a shared library.
+A CHERIoT RTOS shared library has code and read-only globals but no read-write globals and does not require crossing a security boundary to invoke.
+Calls to shared libraries are about as fast as calls to functions in the current compartment but the same code can be shared by many compartments.
+The message-queue library avoids any cross-compartment calls on the fast path and calls the scheduler only when a producer or consumer needs to sleep (when the queue is full or empty, respectively).
+
+Often, CHERIoT RTOS systems will use message queues for communicating between threads in the same compartment and so this is sufficient.
+Alternatively, you may want to use them in a way that's closer to traditional inter-process communication, where the sender and receiver are in different compartments and a shared library with both ends having complete access to the queue would be a problem.
+For this use case, we provide a compartment that owns the message queue and provides sealed capabilities as opaque handles for the endpoints.
+
+## CHERI doesn't look like an MPU
+
+A lot of embedded operating systems were not designed with any kind of protection in mind and had some MPU support retrofitted.
+Others were designed to take advantage of an MPU.
+Neither ends up with abstractions that map well to CHERI.
+
+For example, FreeRTOS has the notion of a *restricted task*, which is similar to a process in a conventional operating system.
+A restricted task has access to a limited set of MPU regions, rather than the entire address space.
+This model lets you run some code with reduced privileges, but does not help with the *principle of intentional use* and so does not align well with a CHERI model.
+
+In an MPU-based approach, you may wish a component to have access to (the memory ranges containing) objects A and B, so you would configure MPU regions that grant access to A and B.
+If there is a bounds error in addressing into A then the component may accidentally overwrite B.
+In contrast, in a CHERI system, you would pass bounded pointers (CHERI capabilities) to A and B.
+Any operation that attempts to access A will never accidentally access B.
+This intentionality (you can access an object only when you intend to access *that object*) is core to CHERI.
+
+The CHERIoT security model was designed to support language-friendly abstractions.
+You don't share objects in CHERIoT RTOS by creating shared regions of memory and populating them, you share objects by passing pointers as arguments to functions.
```

Comment:
❤️ 

I don't mind stating this twenty more times to hammer this point. This is basically going to be huge a performance win.